src/Storages/StorageS3.cpp

```diff
@@ -1017,13 +1069,18 @@ Pipe StorageS3::read(
     std::shared_ptr<StorageS3Source::IIterator> iterator_wrapper = createFileIterator(
         query_configuration, distributed_processing, local_context, query_info.query, virtual_columns, nullptr, local_context->getFileProgressCallback());
 
+    size_t estimated_keys_count = iterator_wrapper->estimatedKeysCount();
+    num_streams = std::min(num_streams, estimated_keys_count);
+    LOG_INFO(&Poco::Logger::get("StorageS3"), "adjusting num_streams={}", num_streams);
```

Comment:
Nitpick: Info seems too strong for this, should it be debug?
Nitpick 2: Maybe say something like "Reading in {} streams, {} threads per stream" instead, since (a) it's printed even if num_streams wasn't adjusted, (b) maximize useful information per log line.